libp2p/host.ss

```diff
@@ -0,0 +1,78 @@
+;;; -*- Gerbil -*-
+;;; libp2p host implementation
+
+(import :std/io
+        :std/misc/timeout
+        :std/format
+        :std/sugar
+        :std/crypto
+        :std/text/base58
+        :std/logger
+        "utils"
+        "multiaddress")
+
+;; TODO This should derive the peer-id from the key-pair
+(def (make-peer-id key)
+  (base58-encode key))
+
+(defstruct peer-info (peer-id multi-address public-key))
+
+(defstruct p2p-host (id key multi-address listener peer-store handlers connections)
+  constructor: init!)
+
+(defmethod {init! p2p-host}
+  (lambda (self (key (keygen/ed25519)) (multi-addr "ip4/127.0.0.1/tcp/4444"))
+    (set! (@ self multi-address) (multi-address multi-addr))
+    (set! (@ self peer-store) (make-hash-table))
+    (set! (@ self handlers) (make-hash-table))
+    (set! (@ self connections) (make-hash-table))
+    ;; TODO Make this a proper implementation
+    (set! (@ self id) (make-peer-id key))))
+
+(def (tcp-listener address (proc 123))
```

Comment:
use type annotations here for the arguments, as the can be checked at runtime for you at the call boundary, and often times be eliminated (the checks) when the compiler knows the type.